README.md
# Progress Bar for PHP CLI Scripts / Shell Environment

This project aims to deliver a easy to use php component for displaying a progress bar in the shell environment.

The current change log can be found [here](https://github.com/bazzline/php_component_apache_status_parser/blob/master/CHANGELOG.md).

## Usage

```php
//assuming $items contains an array of items
$progressBar = new \Net\Bazzline\Component\Cli\ProgressBar\ProgressBar();

$progressBar->setTotalSteps(count($items));

foreach ($items as $item) {
    //do something with the item
    $progressBar->forward();
}
//to be on the save side, tell the progress bar that the show is over ;-)
$progressBar->isFinished();
```

## Example

```
cd <project root>
./example/run.php

+-----------------------------+
| Example of the progress bar |
+-----------------------------+
This demo will demonstrate 5 progress bars.
The number of items will be 3, 6, 9, 12, 150.
The current update interval is 1 second.

Press CTRL-C to stop this demonstration

[========================================================================>] (3/3)
```

# Links (other good projects)

* https://github.com/levi-putna/phpcli
* https://github.com/guiguiboy/PHP-CLI-Progress-Bar
* https://nixshell.wordpress.com/2009/09/08/get-the-width-of-the-terminal/

# Final Words

Star it if you like it :-). Add issues if you need it. Pull patches if you enjoy it. Write a blog entry if you use it :-D.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Open]

### To Add

* add link to openhub
* add link to scrutinizer.yml
* add link to packagist (and install howto)
* add api documentation
* add formatter
* add percentage at the beginning that gets calculated and updated on each update call (like wget is doing it)
* add support to calculate maximum width (based on terminal size) dynamically (`tput cols`?)
* add [eta calculation](https://github.com/guiguiboy/PHP-CLI-Progress-Bar/issues/4)
* write [tests](https://github.com/guiguiboy/PHP-CLI-Progress-Bar/blob/master/test/ProgressBar/Test/ManagerTest.php)

### To Change

## [Unreleased]

### Added

### Changed

## [1.1.0](https://github.com/bazzline/php_component_cli_progress_bar/tree/1.1.0) - released at 2017-05-21

### Changed

* moved to psr-4 autoloading
* moved history entry to dedicated changelog file
* updated minimum version of php to 5.6

## [1.0.0](https://github.com/bazzline/php_component_cli_progress_bar/tree/1.0.0) - released at 2015-05-24

### Added

* initial release


source/Net/Bazzline/Component/Cli/ProgressBar/ProgressBar.php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2015-05-24 
 */

namespace Net\Bazzline\Component\Cli\ProgressBar;

class ProgressBar
{
    /** @var int */
    private $currentStep = 0;

    /** @var bool */
    private $initialStepWasZero = false;

    /** @var bool */
    private $isFinished = true;

    /** @var int */
    private $totalStepSize;

    /** @var int */
    private $totalSteps;

    /**
     * @param int $total
     */
    public function setTotalSteps($total)
    {
        $this->totalSteps   = (int) $total;
        $this->calculateNumberOfSteps();
        $this->currentStep  = 0;
        $this->isFinished   = false;
        $this->storeCursorPosition();
    }

    public function isFinished()
    {
        $this->update($this->totalSteps);
    }

    /**
     * @param int $current
     */
    public function update($current)
    {
        if (!$this->isFinished) {
            $this->currentStep = (int) $current;

            if ($this->currentStep === 0) {
                $this->initialStepWasZero = true;
            }
            if ($this->initialStepWasZero) {
                ++$this->currentStep;
            }

            $this->resetCursor();
            $this->draw();
        }
    }

    public function forward()
    {
        $next = ($this->initialStepWasZero) ? $this->currentStep : ++$this->currentStep;

        $this->update($next);
    }

    private function draw()
    {
        $numberOfStepDraws                  = (($this->currentStep * $this->totalStepSize) / $this->totalSteps);
        $numberOfMissingStepDraws           = ($this->totalStepSize - $numberOfStepDraws);
        $numberOfWhiteSpacesForCurrentStep  = (strlen($this->totalSteps) - strlen($this->currentStep));

        $output = '[';
        $output .= str_repeat('=', floor($numberOfStepDraws));
        $output .= '>';
        $output .= str_repeat(' ', ceil($numberOfMissingStepDraws));
        $output .= '] (';

        if ($numberOfWhiteSpacesForCurrentStep > 0) {
            $output .= str_repeat(' ', $numberOfWhiteSpacesForCurrentStep);
        }
        $output .= $this->currentStep . '/' . $this->totalSteps . ')';

        if ($this->currentStep >= $this->totalSteps) {
            $this->isFinished   = true;
            $output            .= PHP_EOL;
        }

        echo $output;
    }

    private function resetCursor()
    {
        echo "\0338";
    }

    private function storeCursorPosition()
    {
        echo "\0337";
    }

    private function calculateNumberOfSteps()
    {
        $lengthOfTotal = strlen($this->totalSteps);
        $lengthOfAdditionalCharacters = 6; //[] (/)

        $this->totalStepSize = 80 - (2 * $lengthOfTotal) - $lengthOfAdditionalCharacters;
    }
}

example/run.php
#!/bin/php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2015-05-24 
 */

require_once __DIR__ . '/../vendor/autoload.php';

$numberOfItems = [
    3,6,9,12,150
];

echo '+-----------------------------+' . PHP_EOL;
echo '| Example of the progress bar |' . PHP_EOL;
echo '+-----------------------------+' . PHP_EOL;
echo 'This demo will demonstrate ' . count($numberOfItems) . ' progress bars.' . PHP_EOL;
echo 'The number of items will be ' . implode(', ', $numberOfItems) . '.' . PHP_EOL;
echo 'The current update interval is 1 second.' . PHP_EOL;
echo PHP_EOL;
echo 'Press CTRL-C to stop this demonstration.' . PHP_EOL;
echo PHP_EOL;

$progressBar = new \Net\Bazzline\Component\Cli\ProgressBar\ProgressBar();

foreach ($numberOfItems as $totalNumberOfItems) {
    $items      = [];
    $startTime  = microtime(true);

    for ($iterator = 0; $iterator < $totalNumberOfItems; ++$iterator) {
        $items[] = $iterator;
    }
    $totalNumberOfItems = count($items);

    $progressBar->setTotalSteps($totalNumberOfItems);

    foreach ($items as $key => $item) {
        //$progressBar->update($key);
        $progressBar->forward();
        sleep(1);
    }

    $totalTime = microtime(true) - $startTime;

    echo PHP_EOL;
    echo 'total number of items: ' . $totalNumberOfItems . PHP_EOL;
    echo 'total time: ' . (int) $totalTime . ' seconds' . PHP_EOL;
}
